phep-proposal.md

```diff
@@ -0,0 +1,166 @@
+```
+PhEP {
+	#id: 000x,
+	#type: #image,
+	#title: 'New Finalization',
+	#authors: [ 'Guillermo Polito' ],
+	#created: '2022-01-09'
+}
+```
+
+# Abstract 
+
+A proposal describing a new finalization mechanism for Pharo with the following properties:
+
+ - API independent from its implementation
+ - Avoids memory leaks
+
+This document describes both a finalization API as a Pharo library and an efficient virtual machine implementation through Ephemerons.
+
+# Motivation
+
+Some applications require tracking when objects are *about to* be garbage collected to perform some treatment. For example, file descriptors and sockets must be closed when garbage-collected.
+Such a mechanism is implemented through automatic finalization.
+
+The finalization mechanism present in Pharo up to version 10 (inclusive) is suboptimal.
+First, it requires a full scan of many data structures to detect finalization candidates.
+Second, the bad usage of the finalization mechanism leads to memory leaks.
+Third, the current API is tied to its implementation (`WeakRegistry`).
+
+# Description
+
+## Terminology
+
+- **Finalizable**: an object of interest regarding garbage collection. When this object is about to be garbage-collected we want to notify a **finalizer**.
+- **Finalizer**: an object that gets notified when a **finalizable** is about to be garbage-collected to perform some action.
+- **Finalization Registry**: a registry of *(finalizable, finalizer)* pairs containing finalizable objects and their corresponding finalizer.
+
+## Finalization Overview and Protocol
+
+Finalization is managed at the user level by a finalization registry.
+The finalization registry contains (finalizable, finalizer) pairs.
+The registry has two main methods: `#add:` and `#add:finalizer:`.
+Finalizers need to implement `#finalize` with their corresponding finalization
+
+- `FinalizationRegistry>>#add:` adds an object as both finalizable and finalizer. This means the object is its own finalizer.
+- `FinalizationRegistry>>#add:finalizer:` adds an object as finalizable with another object as finalizer.
+- `[User Object]>>#finalize` invokes the finalization mechanism of the given finalizer. The default `#finalize` method in `Object` is defined to do nothing. 
```

Comment:
To be consistent with the lack of constraints on multiplicity, I would have a parameter here that is the finalizable object given with `add:finalizer:` (or self if plain `add:` is used). ie `UserFinaliserClass>>#finalize: aUserFinalizableObject`